rcp/org.testeditor.rcp4.views.teststepselection/src/org/testeditor/rcp4/views/teststepselector/TestStepSelector.xtend

```diff
@@ -110,39 +148,87 @@ class TestStepSelector {
 //		// startup complete ensures the index to be populated
 //		Display.getDefault.syncExec[populateViewIfEmpty] // is only run, if view is created on startup
 //	}
-
 	@Focus
 	def void setFocus() {
 		viewer.control.setFocus
 	}
 
 	@Inject
 	@Optional
+	def void updateView(@EventTopic(SELECTOR_TOPIC_UPDATE) Object data) {
+		logger.debug("updateView.")
+		Display.getDefault.syncExec[updateView(true)]
+	}
+
+	@Inject
+	@Optional
 	def void refreshView(@EventTopic(SELECTOR_TOPIC_REFRESH) Object data) {
-		populated = false
-		logger.debug("Refreshing view.")
-		Display.getDefault.syncExec[populateViewIfEmpty]
+		logger.debug("refreshView.")
+		Display.getDefault.syncExec[updateView(false)]
 	}
 
-	// make sure that the index is populated before calling this method
-	// and that it is run in the swt thread!
-	private def populateViewIfEmpty() {
-		if (populated) {
+	private def updateView(boolean update) {
+
+		// check if a update is needed
+		if (!update && viewer.input != null) {
 			return
 		}
+
+		// update model
 		val amlInjector = amlInjectorProvider.get
+		val model = amlInjector.getInstance(AmlModelsProvider).amlModels
+
+		// update view and set model
+		if (viewer.input == null) {
+			viewer.labelProvider = labelProvider
+			// TODO different handling of labelProvider and contentProvider
+			viewer.contentProvider = amlInjectorProvider.get.getInstance(TestStepSelectorTreeContentProvider)
+			viewer.input = model
+		} else {
+			val Set<String> expandedElements = viewer.expandedElements.map[toStringPath(it)].toSet
+			viewer.input = model;
+			viewer.expandedElements = expandElements(expandedElements, model)
+		}
 
-		populated = true
-		viewer.labelProvider = labelProvider
-		viewer.contentProvider = amlInjector.getInstance(TestStepSelectorTreeContentProvider)
+	}
 
-		val amlModelsProvider = amlInjector.getInstance(AmlModelsProvider)
-		viewer.input = amlModelsProvider.amlModels
-		viewer.expandAll
+	private def Object[] expandElements(Set<String> elements, Iterable<AmlModel> model) {
+		val elementsToExpand = new ArrayList<Object>();
+		model.forEach [
+			{
+				if (elements.contains(it.toStringPath)) {
+					elementsToExpand.add(it.toStringPath)
```

Comment:
This should be done using a `map` method as done with the initial retrieval of the expanded elements.